### PR TITLE
Implement vehicle order table reversing

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
Implements #1972

Screenshot Screenshot (1)

### CODE REVIEW

File: src/OpenLoco/src/Vehicles/Orders.cpp [hunk 1]
```diff
@@ -206,6 +206,68 @@ namespace OpenLoco::Vehicles
...
+            std::memcpy(dest, &rawOrder, orderLength);
+            dest += orderLength;
+        }
+
+        printf("First order @ %p\n", firstOrder);
+        printf("End of table @ %p\n", dest);
+
+        // Figure out the new position of the order of interest
+        auto newOOIOffset = dest - orderOfInterest - kOrderSizes[enumValue(ooiType)];
+        printf("Calculated OOI @ %p\n", newOOIOffset);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Vehicles/Orders.cpp:266]: This does not appear to be yielding the correct offset at present. Not sure what the problem is yet.

File: src/OpenLoco/src/Windows/Vehicle.cpp [hunk 2]
```diff
@@ -211,6 +212,7 @@ namespace OpenLoco::Ui::Windows::Vehicle
             makeWidget({ 240, 116 }, { 24, 24 }, WidgetType::buttonWithImage, WindowColour::secondary, ImageIds::route_delete, StringIds::tooltip_route_delete_order),
             makeWidget({ 240, 140 }, { 24, 12 }, WidgetType::buttonWithImage, WindowColour::secondary, ImageIds::red_arrow_up, StringIds::tooltip_route_move_order_up),
             makeWidget({ 240, 152 }, { 24, 12 }, WidgetType::buttonWithImage, WindowColour::secondary, ImageIds::red_arrow_down, StringIds::tooltip_route_move_order_down),
+            makeWidget({ 240, 164 }, { 24, 24 }, WidgetType::buttonWithImage, WindowColour::secondary, ImageIds::construction_right_turnaround),
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Windows/Vehicle.cpp:215]: Should add a tooltip for this new widget.

File: src/OpenLoco/src/Vehicles/Orders.cpp [hunk 3]
```diff
@@ -206,6 +206,45 @@ namespace OpenLoco::Vehicles
         return begin();
     }
 
+    uint16_t reverseVehicleOrderTable(uint16_t tableOffset, uint16_t orderOfInterest)
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Vehicles/Orders.cpp:209]: Probably should go into `OrderManager`, if we move `kOrderSizes` there as well.

**Review by LeftofZen — APPROVED** [review]
playtested this, seemed to work with no issues for me. it would be nice to preserve the ordering of the pickup-dropoff commands with destionations, but as discussed already that would require a lot more work, so this implementation is good for now

**Comment by AaronVanGeffen** [comment]
@LeftofZen Thanks for testing. Yes, @YetiWiz's request is sensible, but a bunch more work with the current data structures.

Before merging this, I'd like to move `reverseVehicleOrderTable` to `OrderManager`. That becomes feasible after #1978 has been merged.

**Comment by AaronVanGeffen** [comment]
Now that #1978 has been merged, I have rebased this PR and moved `reverseVehicleOrderTable` into `OrderManager`. This is now good to go, as far as I'm concerned.

File: src/OpenLoco/src/Vehicles/OrderManager.cpp [hunk 4]
```diff
@@ -280,6 +280,45 @@ namespace OpenLoco::Vehicles::OrderManager
         }
     }
 
+    uint16_t reverseVehicleOrderTable(uint16_t tableOffset, uint16_t orderOfInterest)
+    {
+        // Retrieve list of raw orders
+        std::vector<uint64_t> rawOrders{};
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Vehicles/OrderManager.cpp:286]: We should probably have a note somewhere in the order manager that you can't use std:: algorithms for a lot of this due to the variable length of elements. Then hopefully when we stop using variable length we can simplify a huge amount of this.

File: src/OpenLoco/src/Windows/Vehicle.cpp [hunk 5]
```diff
@@ -193,12 +193,13 @@ namespace OpenLoco::Ui::Windows::Vehicle
...
             orderUp,
-            orderDown
+            orderDown,
+            orderReverse
         };
 
         // 0x00500554
         static WindowEventList events;
-        constexpr uint64_t enabledWidgets = (1 << widx::routeList) | (1 << widx::orderForceUnload) | (1 << widx::orderWait) | (1 << widx::orderSkip) | (1 << widx::orderDelete) | (1 << widx::orderUp) | (1 << widx::orderDown) | Common::enabledWidgets;
+        constexpr uint64_t enabledWidgets = (1 << widx::routeList) | (1 << widx::orderForceUnload) | (1 << widx::orderWait) | (1 << widx::orderSkip) | (1 << widx::orderDelete) | (1 << widx::orderUp) | (1 << widx::orderDown) | (1 << widx::orderReverse) | Common::enabledWidgets;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Windows/Vehicle.cpp:202]: should really be using 1ULL for constructing a uint64_t. I know we don't do that consistently through the codebase though.

File: src/OpenLoco/src/Windows/Vehicle.cpp [hunk 6]
```diff
@@ -2480,6 +2482,18 @@ namespace OpenLoco::Ui::Windows::Vehicle
...
 
+        static bool orderReverseCommand(Vehicles::VehicleHead* const head)
+        {
+            GameCommands::VehicleOrderReverseArgs args{};
+            args.head = head->id;
+
+            GameCommands::setErrorTitle(StringIds::empty);
+            auto result = GameCommands::doCommand(args, GameCommands::Flags::apply);
+
+            Vehicles::OrderManager::generateNumDisplayFrames(head); // Note: order changed, check if this matters.
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Windows/Vehicle.cpp:2493]: note is not relevant for this

File: data/language/en-GB.yml [hunk 7]
```diff
@@ -2334,3 +2334,4 @@ strings:
   2279: "{COLOUR BLACK}Checksum"
   2280: "Play sound effects for new messages"
   2281: "{SMALLFONT}{COLOUR BLACK}When enabled, new messages will be accompanied by a sound effect of an appropriately cheering crowd."
+  2282: "{SMALLFONT}{COLOUR BLACK}Reverse the route order table"
```
Comment (duncanspumpkin) [inline data/language/en-GB.yml:2337]: is it called order table in game?

File: src/OpenLoco/src/Vehicles/OrderManager.cpp (hunk 4 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Vehicles/OrderManager.cpp:286]: Completely agreed. I'll add the note in a few places. I'm thinking something among the lines of:
```
// Currently the use of std:: algorithms is not feasible due to variable order lengths
// TODO: simplify this after the changing the data structure for the order table
```

File: data/language/en-GB.yml (hunk 7 above)
Comment (AaronVanGeffen) [reply data/language/en-GB.yml:2337]: I hesitated to call it such as well. The tab is actually called 'Routing', but the command tooltips refer to 'orders'. As a compromise, I thought I'd refer to both in this way.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**